exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt

```diff
@@ -1061,30 +1064,61 @@ abstract class VendorDialect(
         resetCaches()
     }
 
+    fun filterCondition(index: Index): String? {
+        return if (currentDialect is PostgreSQLDialect) {
+            index.filterCondition?.let {
+                QueryBuilder(false)
+                    .append(" WHERE ").append(it)
+                    .toString()
+            }
+        } else {
+            null
+        }
+    }
+
+    /**
+     * The uniquieness might be required for foreign constraints
+     *
+     * In postgresq (https://www.postgresql.org/docs/current/indexes-unique.html) Uniq means btree only.
+     * Unique constraints can not be partial
+     * Unique indexes can be partial
+     */
     override fun createIndex(index: Index): String {
         val t = TransactionManager.current()
         val quotedTableName = t.identity(index.table)
         val quotedIndexName = t.db.identifierManager.cutIfNecessaryAndQuote(index.indexName)
         val columnsList = index.columns.joinToString(prefix = "(", postfix = ")") { t.identity(it) }
+
+        val maybeFilterCondition = filterCondition(index) ?: ""
+
         return when {
-            index.unique -> {
+            // uniq and no filter -> constraint, the type is not supported
+            index.unique && maybeFilterCondition.isEmpty() -> {
                 "ALTER TABLE $quotedTableName ADD CONSTRAINT $quotedIndexName UNIQUE $columnsList"
             }
+            // uniq and filter -> index only, the type is not supported
```

Comment:
nit (for consistency across the codebase):
```suggestion
            // unique and filter -> index only, the type is not supported
```